clients/src/main/java/org/apache/kafka/common/security/auth/Authorizer.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
+ * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
+ * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.common.security.auth;
+
+import org.apache.kafka.common.Configurable;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Top level interface that all plugable authorizer must implement. Kafka server will read "authorizer.class" config
+ * value at startup time, create an instance of the specified class and call initialize method.
+ * authorizer.class must be a class that implements this interface.
+ * If authorizer.class has no value specified no authorization will be performed.
+ *
+ * From that point onwards, every client request will first be routed to authorize method and the request will only be
+ * authorized if the method returns true.
+ */
+public interface Authorizer extends Configurable {
+
+    /**
+     * @param session   The session being authenticated.
+     * @param operation Type of operation client is trying to perform on resource.
+     * @param resource  Resource the client is trying to access.
+     * @return
+     *
+     * @throws org.apache.kafka.common.errors.InvalidResourceException if resource does not exist
```

Comment:
I think just say "if the resource type does not exist" to constrain it the . This might need more clarification in Javadocs for users implementing this since "resource does not exist" seems to require returning false, but `InvalidResourceException` really means invalid resource *type* (but someone could easily interpret "invalid resource" to include non-existant but correctly-defined resources).